[WIP] Introduce `OS.set_drag_mode()` - Tell the window manager to move / resize the window.

JoshTheDerf opened the pull request:
Borderless windows have a big disadvantage compared to normal windows — they can't be dragged (drug?) or resized without implementing a significant amount of custom logic. And even if you do implement said logic, you'll still lose features such as window snapping. This PR introduces `OS.set_drag_mode(DragMode drag_mode)` and related getters and accessors.

The available drag modes are:

```
DRAG_MODE_NONE, // Current / default behavior.
DRAG_MODE_MOVE, // Tell the window manager to move the window when the mouse is dragged.
DRAG_MODE_RESIZE_TOP, // Tells the window manager to resize the window from the top when the mouse is dragged.
DRAG_MODE_RESIZE_RIGHT, // etc.
DRAG_MODE_RESIZE_BOTTOM,
DRAG_MODE_RESIZE_LEFT,
DRAG_MODE_RESIZE_TOPLEFT,
DRAG_MODE_RESIZE_TOPRIGHT,
DRAG_MODE_RESIZE_BOTTOMRIGHT,
DRAG_MODE_RESIZE_BOTTOMLEFT
```

The drag modes may be set at any time. They will only take effect when the mouse is held down and starts moving in the window.

Example Usage:

To make a panel draggable, add this to the panel's script:
```gdscript
func _on_Panel_mouse_entered():
	OS.set_drag_mode(OS.DRAG_MODE_MOVE);

func _on_Panel_mouse_exited():
	OS.set_drag_mode(OS.DRAG_MODE_NONE);
```
This will treat the panel as a titlebar. Buttons or other interactable controls inside the panel will still be usable as normal without triggering a window move.

Enabling window resize on a background panel is a bit trickier. Say you have a root panel with 5 pixels of internal margin. Here's how you'd make it control window resize:

```gdscript
func _on_Root_Panel_gui_input( ev ):
	if ev is InputEventMouseMotion:
		var horizontal = -1;
		var vertical = -1;
		var mode = OS.DRAG_MODE_NONE;
		var cursor = Control.CURSOR_ARROW;

		if ev.position.x > OS.window_size.x - 5:
			horizontal = OS.DRAG_MODE_RESIZE_RIGHT;
			cursor = Control.CURSOR_HSIZE;
			mode = horizontal;

		if ev.position.x < 5:
			cursor = Control.CURSOR_HSIZE;
			horizontal = OS.DRAG_MODE_RESIZE_LEFT;
			mode = horizontal;

		if ev.position.y < 5:
			cursor = Control.CURSOR_VSIZE;
			vertical = OS.DRAG_MODE_RESIZE_TOP;
			mode = vertical;

		if ev.position.y > OS.window_size.y - 5:
			cursor = Control.CURSOR_VSIZE;
			vertical = OS.DRAG_MODE_RESIZE_BOTTOM;
			mode = vertical;

		if vertical == OS.DRAG_MODE_RESIZE_TOP && horizontal == OS.DRAG_MODE_RESIZE_RIGHT:
			mode = OS.DRAG_MODE_RESIZE_TOPRIGHT;
			cursor = Control.CURSOR_BDIAGSIZE;

		elif vertical == OS.DRAG_MODE_RESIZE_TOP && horizontal == OS.DRAG_MODE_RESIZE_LEFT:
			mode = OS.DRAG_MODE_RESIZE_TOPLEFT;
			cursor = Control.CURSOR_FDIAGSIZE;

		elif vertical == OS.DRAG_MODE_RESIZE_BOTTOM && horizontal == OS.DRAG_MODE_RESIZE_RIGHT:
			mode = OS.DRAG_MODE_RESIZE_BOTTOMRIGHT;
			cursor = Control.CURSOR_FDIAGSIZE;

		elif vertical == OS.DRAG_MODE_RESIZE_BOTTOM && horizontal == OS.DRAG_MODE_RESIZE_LEFT:
			mode = OS.DRAG_MODE_RESIZE_BOTTOMLEFT;
			cursor = Control.CURSOR_BDIAGSIZE;

		if OS.drag_mode != mode:
			self.mouse_default_cursor_shape = cursor;
			OS.set_drag_mode(mode);
```

Together, those scripts are sufficient to allow native window movement and resize support as shown in the GIF below (Pardon the recording quality, the jittering when moving isn't present in actual usage.)

![ezgif com-optimize](https://user-images.githubusercontent.com/5861371/36016154-9fccc106-0dac-11e8-9d05-423e6ce130dd.gif)

## Status:
- [x] X11 Implementation
- [x] macOS Implementation
- [x] Windows Implementation **Doesn't support Aero Snap yet.**
- [ ] Haiku Implementation?
- ~~UWP Implementation?~~ Doesn't appear to be possible.

Android, iPhone, and Web support shouldn't be necessary since they don't have a floating window manager system. (At least for the time being.)

*I know I should really discuss the features for these PRs beforehand on IRC or Discord or something, but unfortunately, my timezone appears to be the opposite of pretty much everyone else, so it's faster for me to just write the code and PR it for discussion.*

groud commented:
It's difficult to evaluate the code like that. Could you rebase and apply clang-format ?

Tribex commented:
@groud Working on that now. The issue is that I *did* apply clang-format, and seems to have either used the wrong ruleset or formatted a great many things. I'm now manually re-adding my changes to the master.

JoshTheDerf committed:
Define, implement, and document OS.set_drag_mode(int) on X11.

OS.set_drag_mode() tells the window manager how to handle drag events on your app.
For example, setting DRAG_MODE_MOVE will tell the window manager to start moving your window on drag.
Setting DRAG_MODE_RESIZE_BOTTOMRIGHT will tell the window manger to start resizing your window from the bottom right on drag.

Support for these features is critical for borderless apps which need to support normal window manager interactions.

Tribex commented:
@groud Done. Sorry for the trouble.

Zylann commented:
Aren't borderless windows actually handled by the OS sometimes? On Windows for example I know that borders actually exist and react to input, but are simply hidden. On the other hand, I witnessed coders struggling to reimplement all window manager features on their borderless app and miserably failing to get good UX (drag areas were fucked, the window had no shadow, no drag-on-top-to-maximize, no double-click and all...)

Tribex commented:
@Zylann I believe Godot doesn't use the invisible / overlay titlebar hack. If it does, I'll disable it in favor of this for flexibility purposes. This implementation provides access to window manager features without needing custom implementations, which is my goal.

JoshTheDerf committed:
Initial macOS support. (DRAG_MODE_MOVE only.)

Unfortunately, implementing resize on macOS will likely need to be done manually due to there being fewer APIs to control such features. If anyone knows better, please let me know.

JoshTheDerf committed:
Finish initial OSX support.

Now supports resize and maximize by double-clicking titlebar.

Tribex commented:
Alright, the initial macOS implementation should be working. Had to implement resize support semi-manually, but it works great!

JoshTheDerf committed:
Initial Windows Implementation. Supports all modes.

Unfortunately, the way borderless windows are being created will need
to be adjusted in order to support features such as Aero snap.
It should still be possible to get those working, however.

girng commented:
we can already drag to resize the window pane though, what exactly is this suppose to be for?

girng commented:
>We can't (easily) resize and drag borderless 

oh, borderless, i missed that, my bad.

with that said, however shouldn't the main OS window handle the resize, not gui elements? what you are saying is if a borderless windows is enabled, that functionality isn't there, so using a full rect panel node for the drag is the solution basically?

Tribex commented:
@girng 

> with that said, however shouldn't the main OS window handle the resize, not gui elements? what you are saying is if a borderless windows is enabled, that functionality isn't there, so using a full rect panel node for the drag is the solution basically?

Yes, `OS.set_drag_mode()` tells the OS window to handle resize and move as opposed to handling it inside Godot.
Whether or not a panel is used doesn't matter. That's just the method I'm demonstrating. The goal of this PR is to add a way to tell the OS to natively handle resize and window movement.

> a godot dev would have to set this up, and have it be a full rect node in the background though

Basically, yes, though they could use other methods as well. You could even call it from 3D spatials, though I wouldn't recommend it.

> which is prob already possible w/o this PR, isn't it? using just gdscript w/ _input
Right now it can be emulated, but not sufficiently. On most platforms, the windows will be jumpy and glitchy, and they won't support features such as window snapping (Aero Snap on Windows) and such.

With this PR, developers won't need to try to implement hacky custom move / resize code. They'll just need to make a call to `OS.set_drag_mode()` when the cursor enters and exits an element that should trigger native OS move / resize.

The reason the API is a bit odd, (setting a mode instead of directly telling the OS window manager to start dragging or resizing) is because on Windows and macOS, the window must know whether or not it can start dragging before it will attempt to.

Tribex commented:
Aaaand it's working on Windows now. Well, mostly. Aero Snap doesn't function yet due to the way borderless windows are created. I'll probably need to reimplement borderless windows in a different PR.

DuskKittyIsHotAF commented:
very tolerant and nice commnuity imo. run by great leadership skills, you should be proud juan and akien for leading this great floss project!
![](https://i.gyazo.com/2fd0c035013b3c1bb0652552d35fde92.png)
![](https://i.gyazo.com/e0acd22258872330f342d5d189564f1e.png)
![](https://i.gyazo.com/81b238ce283aa55126e3a2f608583837.png)

DuskKittyIsHotAF commented:
very tolerant and nice commnuity imo. run by great leadership skills, you should be proud juan and akien for leading this great floss project!
![](https://i.gyazo.com/2fd0c035013b3c1bb0652552d35fde92.png)
![](https://i.gyazo.com/e0acd22258872330f342d5d189564f1e.png)
![](https://i.gyazo.com/81b238ce283aa55126e3a2f608583837.png)

DuskKittyIsHotAF commented:
very tolerant and nice commnuity imo. run by great leadership skills, you should be proud juan and akien for leading this great floss project!
![](https://i.gyazo.com/2fd0c035013b3c1bb0652552d35fde92.png)
![](https://i.gyazo.com/e0acd22258872330f342d5d189564f1e.png)
![](https://i.gyazo.com/81b238ce283aa55126e3a2f608583837.png)

DuskKittyIsHotAF commented:
very tolerant and nice commnuity imo. run by great leadership skills, you should be proud juan and akien for leading this great floss project!
![](https://i.gyazo.com/2fd0c035013b3c1bb0652552d35fde92.png)
![](https://i.gyazo.com/e0acd22258872330f342d5d189564f1e.png)
![](https://i.gyazo.com/81b238ce283aa55126e3a2f608583837.png)

reduz commented:
if this works, It's ok I guess, @TMM, @BastiaanOlij, @vnen  ?
seems check fails also?

Tribex commented:
Might need to re-add. Not sure what the failing check is.

Noshyaar commented:
Styling issues. Please run clang-format and squash/amend the commit.

<details>

```
--- a/core/bind/core_bind.h	2018-02-10 13:18:32.822590095 +0000
+++ b/core/bind/core_bind.h	2018-02-10 13:19:29.092864595 +0000
@@ -137,22 +137,22 @@
 	void set_window_title(const String &p_title);
 	int get_mouse_button_state() const;
 
-  enum DragMode {
+	enum DragMode {
 
-    DRAG_MODE_NONE,
-    DRAG_MODE_MOVE,
-    DRAG_MODE_RESIZE_TOP,
-    DRAG_MODE_RESIZE_RIGHT,
-    DRAG_MODE_RESIZE_BOTTOM,
-    DRAG_MODE_RESIZE_LEFT,
-    DRAG_MODE_RESIZE_TOPLEFT,
-    DRAG_MODE_RESIZE_TOPRIGHT,
-    DRAG_MODE_RESIZE_BOTTOMRIGHT,
-    DRAG_MODE_RESIZE_BOTTOMLEFT
-  };
+		DRAG_MODE_NONE,
+		DRAG_MODE_MOVE,
+		DRAG_MODE_RESIZE_TOP,
+		DRAG_MODE_RESIZE_RIGHT,
+		DRAG_MODE_RESIZE_BOTTOM,
+		DRAG_MODE_RESIZE_LEFT,
+		DRAG_MODE_RESIZE_TOPLEFT,
+		DRAG_MODE_RESIZE_TOPRIGHT,
+		DRAG_MODE_RESIZE_BOTTOMRIGHT,
+		DRAG_MODE_RESIZE_BOTTOMLEFT
+	};
 
-  void set_drag_mode(DragMode p_drag_mode);
-  DragMode get_drag_mode() const;
+	void set_drag_mode(DragMode p_drag_mode);
+	DragMode get_drag_mode() const;
 
 	void set_clipboard(const String &p_text);
 	String get_clipboard() const;
--- a/core/os/os.cpp	2018-02-10 13:18:32.822590095 +0000
+++ b/core/os/os.cpp	2018-02-10 13:19:29.116701883 +0000
@@ -525,7 +525,9 @@
 	return MOUSE_MODE_VISIBLE;
 }
 
-OS::DragMode OS::get_drag_mode() const { return DRAG_MODE_NONE; }
+OS::DragMode OS::get_drag_mode() const {
+	return DRAG_MODE_NONE;
+}
 
 OS::LatinKeyboardVariant OS::get_latin_keyboard_variant() const {
 
--- a/platform/osx/os_osx.mm	2018-02-10 13:18:32.826589740 +0000
+++ b/platform/osx/os_osx.mm	2018-02-10 13:19:29.193064156 +0000
@@ -556,43 +556,43 @@
 	switch (OS_OSX::singleton->drag_mode) {
 		case OS::DRAG_MODE_RESIZE_TOP:
 			[[event window] setFrame:NSMakeRect(
-				frame.origin.x, frame.origin.y, frame.size.width, frame.size.height - event.deltaY
-			) display:YES];
+											 frame.origin.x, frame.origin.y, frame.size.width, frame.size.height - event.deltaY)
+							 display:YES];
 			break;
 		case OS::DRAG_MODE_RESIZE_BOTTOM:
 			[[event window] setFrame:NSMakeRect(
-				frame.origin.x, frame.origin.y - event.deltaY, frame.size.width, frame.size.height + event.deltaY
-			) display:YES];
+											 frame.origin.x, frame.origin.y - event.deltaY, frame.size.width, frame.size.height + event.deltaY)
+							 display:YES];
 			break;
 		case OS::DRAG_MODE_RESIZE_LEFT:
 			[[event window] setFrame:NSMakeRect(
-				frame.origin.x + event.deltaX, frame.origin.y, frame.size.width - event.deltaX, frame.size.height
-			) display:YES];
+											 frame.origin.x + event.deltaX, frame.origin.y, frame.size.width - event.deltaX, frame.size.height)
+							 display:YES];
 			break;
 		case OS::DRAG_MODE_RESIZE_RIGHT:
 			[[event window] setFrame:NSMakeRect(
-				frame.origin.x, frame.origin.y, frame.size.width + event.deltaX, frame.size.height
-			) display:YES];
+											 frame.origin.x, frame.origin.y, frame.size.width + event.deltaX, frame.size.height)
+							 display:YES];
 			break;
 		case OS::DRAG_MODE_RESIZE_TOPLEFT:
 			[[event window] setFrame:NSMakeRect(
-				frame.origin.x + event.deltaX, frame.origin.y, frame.size.width - event.deltaX, frame.size.height - event.deltaY
-			) display:YES];
+											 frame.origin.x + event.deltaX, frame.origin.y, frame.size.width - event.deltaX, frame.size.height - event.deltaY)
+							 display:YES];
 			break;
 		case OS::DRAG_MODE_RESIZE_BOTTOMLEFT:
 			[[event window] setFrame:NSMakeRect(
-				frame.origin.x + event.deltaX, frame.origin.y - event.deltaY, frame.size.width - event.deltaX, frame.size.height + event.deltaY
-			) display:YES];
+											 frame.origin.x + event.deltaX, frame.origin.y - event.deltaY, frame.size.width - event.deltaX, frame.size.height + event.deltaY)
+							 display:YES];
 			break;
 		case OS::DRAG_MODE_RESIZE_TOPRIGHT:
 			[[event window] setFrame:NSMakeRect(
-				frame.origin.x, frame.origin.y, frame.size.width + event.deltaX, frame.size.height - event.deltaY
-			) display:YES];
+											 frame.origin.x, frame.origin.y, frame.size.width + event.deltaX, frame.size.height - event.deltaY)
+							 display:YES];
 			break;
 		case OS::DRAG_MODE_RESIZE_BOTTOMRIGHT:
 			[[event window] setFrame:NSMakeRect(
-				frame.origin.x, frame.origin.y - event.deltaY, frame.size.width + event.deltaX, frame.size.height + event.deltaY
-			) display:YES];
+											 frame.origin.x, frame.origin.y - event.deltaY, frame.size.width + event.deltaX, frame.size.height + event.deltaY)
+							 display:YES];
 			break;
 		default:
 			[self mouseMoved:event];
@@ -715,7 +715,7 @@
 	trackingArea = [[NSTrackingArea alloc]
 			initWithRect:[self bounds]
 				 options:options
-					 owner:self
+				   owner:self
 				userInfo:nil];
 
 	[self addTrackingArea:trackingArea];
@@ -1097,9 +1097,9 @@
 
 	window_object = [[GodotWindow alloc]
 			initWithContentRect:NSMakeRect(0, 0, p_desired.width, p_desired.height)
-						styleMask:styleMask
+					  styleMask:styleMask
 						backing:NSBackingStoreBuffered
-							defer:NO];
+						  defer:NO];
 
 	ERR_FAIL_COND_V(window_object == nil, ERR_UNAVAILABLE);
 
@@ -1130,7 +1130,9 @@
 	[window_object center];
 
 	[window_object setRestorable:NO];
-	if (p_desired.borderless_window) { [window_object setHasShadow:YES]; }
+	if (p_desired.borderless_window) {
+		[window_object setHasShadow:YES];
+	}
 
 	unsigned int attributeCount = 0;
 
@@ -1420,13 +1422,13 @@
 
 		NSBitmapImageRep *imgrep = [[[NSBitmapImageRep alloc]
 				initWithBitmapDataPlanes:NULL
-								pixelsWide:image->get_width()
-								pixelsHigh:image->get_height()
-							 bitsPerSample:8
+							  pixelsWide:image->get_width()
+							  pixelsHigh:image->get_height()
+						   bitsPerSample:8
 						 samplesPerPixel:4
 								hasAlpha:YES
 								isPlanar:NO
-							colorSpaceName:NSDeviceRGBColorSpace
+						  colorSpaceName:NSDeviceRGBColorSpace
 							 bytesPerRow:image->get_width() * 4
 							bitsPerPixel:32] autorelease];
 
@@ -1518,13 +1520,13 @@
 	img->convert(Image::FORMAT_RGBA8);
 	NSBitmapImageRep *imgrep = [[[NSBitmapImageRep alloc]
 			initWithBitmapDataPlanes:NULL
-							pixelsWide:img->get_width()
-							pixelsHigh:img->get_height()
-						 bitsPerSample:8
+						  pixelsWide:img->get_width()
+						  pixelsHigh:img->get_height()
+					   bitsPerSample:8
 					 samplesPerPixel:4
 							hasAlpha:YES
 							isPlanar:NO
-						colorSpaceName:NSDeviceRGBColorSpace
+					  colorSpaceName:NSDeviceRGBColorSpace
 						 bytesPerRow:img->get_width() * 4
 						bitsPerPixel:32] autorelease];
 	ERR_FAIL_COND(imgrep == nil);
@@ -1644,7 +1646,7 @@
 	NSPasteboard *pasteboard = [NSPasteboard generalPasteboard];
 	[pasteboard declareTypes:types owner:nil];
 	[pasteboard setString:[NSString stringWithUTF8String:p_text.utf8().get_data()]
-					forType:NSStringPboardType];
+				  forType:NSStringPboardType];
 }
 
 String OS_OSX::get_clipboard() const {
@@ -2142,8 +2144,8 @@
 		NSEvent *event = [NSApp
 				nextEventMatchingMask:NSEventMaskAny
 							untilDate:[NSDate distantPast]
-								 inMode:NSDefaultRunLoopMode
-								dequeue:YES];
+							   inMode:NSDefaultRunLoopMode
+							  dequeue:YES];
 
 		if (event == nil)
 			break;
@@ -2426,8 +2428,8 @@
 		NSEvent *event = [NSApp
 				nextEventMatchingMask:NSEventMaskAny
 							untilDate:[NSDate distantPast]
-								 inMode:NSDefaultRunLoopMode
-								dequeue:YES];
+							   inMode:NSDefaultRunLoopMode
+							  dequeue:YES];
 
 		if (event == nil)
 			break;
--- a/platform/x11/os_x11.cpp	2018-02-10 13:18:32.826589740 +0000
+++ b/platform/x11/os_x11.cpp	2018-02-10 13:19:29.387039645 +0000
@@ -350,18 +350,18 @@
 	XSetWindowAttributes new_attr;
 
 	new_attr.event_mask = KeyPressMask | KeyReleaseMask | ButtonPressMask |
-							ButtonReleaseMask | EnterWindowMask |
-							LeaveWindowMask | PointerMotionMask |
-							Button1MotionMask |
-							Button2MotionMask | Button3MotionMask |
-							Button4MotionMask | Button5MotionMask |
-							ButtonMotionMask | KeymapStateMask |
-							ExposureMask | VisibilityChangeMask |
-							StructureNotifyMask |
-							SubstructureNotifyMask | SubstructureRedirectMask |
-							FocusChangeMask | PropertyChangeMask |
-							ColormapChangeMask | OwnerGrabButtonMask |
-							im_event_mask;
+						  ButtonReleaseMask | EnterWindowMask |
+						  LeaveWindowMask | PointerMotionMask |
+						  Button1MotionMask |
+						  Button2MotionMask | Button3MotionMask |
+						  Button4MotionMask | Button5MotionMask |
+						  ButtonMotionMask | KeymapStateMask |
+						  ExposureMask | VisibilityChangeMask |
+						  StructureNotifyMask |
+						  SubstructureNotifyMask | SubstructureRedirectMask |
+						  FocusChangeMask | PropertyChangeMask |
+						  ColormapChangeMask | OwnerGrabButtonMask |
+						  im_event_mask;
 
 	XChangeWindowAttributes(x11_display, x11_window, CWEventMask, &new_attr);
 
@@ -2062,35 +2062,35 @@
 	int resize_direction = -1;
 
 	switch (drag_mode) {
-	case DRAG_MODE_MOVE:
-		resize_direction = _NET_WM_MOVERESIZE_MOVE;
-		break;
-	case DRAG_MODE_RESIZE_TOP:
-		resize_direction = _NET_WM_MOVERESIZE_SIZE_TOP;
-		break;
-	case DRAG_MODE_RESIZE_RIGHT:
-		resize_direction = _NET_WM_MOVERESIZE_SIZE_RIGHT;
-		break;
-	case DRAG_MODE_RESIZE_BOTTOM:
-		resize_direction = _NET_WM_MOVERESIZE_SIZE_BOTTOM;
-		break;
-	case DRAG_MODE_RESIZE_LEFT:
-		resize_direction = _NET_WM_MOVERESIZE_SIZE_LEFT;
-		break;
-	case DRAG_MODE_RESIZE_TOPLEFT:
-		resize_direction = _NET_WM_MOVERESIZE_SIZE_TOPLEFT;
-		break;
-	case DRAG_MODE_RESIZE_TOPRIGHT:
-		resize_direction = _NET_WM_MOVERESIZE_SIZE_TOPRIGHT;
-		break;
-	case DRAG_MODE_RESIZE_BOTTOMRIGHT:
-		resize_direction = _NET_WM_MOVERESIZE_SIZE_BOTTOMRIGHT;
-		break;
-	case DRAG_MODE_RESIZE_BOTTOMLEFT:
-		resize_direction = _NET_WM_MOVERESIZE_SIZE_BOTTOMLEFT;
-		break;
-	default:
-		return false;
+		case DRAG_MODE_MOVE:
+			resize_direction = _NET_WM_MOVERESIZE_MOVE;
+			break;
+		case DRAG_MODE_RESIZE_TOP:
+			resize_direction = _NET_WM_MOVERESIZE_SIZE_TOP;
+			break;
+		case DRAG_MODE_RESIZE_RIGHT:
+			resize_direction = _NET_WM_MOVERESIZE_SIZE_RIGHT;
+			break;
+		case DRAG_MODE_RESIZE_BOTTOM:
+			resize_direction = _NET_WM_MOVERESIZE_SIZE_BOTTOM;
+			break;
+		case DRAG_MODE_RESIZE_LEFT:
+			resize_direction = _NET_WM_MOVERESIZE_SIZE_LEFT;
+			break;
+		case DRAG_MODE_RESIZE_TOPLEFT:
+			resize_direction = _NET_WM_MOVERESIZE_SIZE_TOPLEFT;
+			break;
+		case DRAG_MODE_RESIZE_TOPRIGHT:
+			resize_direction = _NET_WM_MOVERESIZE_SIZE_TOPRIGHT;
+			break;
+		case DRAG_MODE_RESIZE_BOTTOMRIGHT:
+			resize_direction = _NET_WM_MOVERESIZE_SIZE_BOTTOMRIGHT;
+			break;
+		case DRAG_MODE_RESIZE_BOTTOMLEFT:
+			resize_direction = _NET_WM_MOVERESIZE_SIZE_BOTTOMLEFT;
+			break;
+		default:
+			return false;
 	}
 
 	XEvent evt;
```

</details>

Timofffee commented:
Bump

Tribex commented:
@Timofffee Sorry, I'm way too swamped at the moment. Might be a while before I can get anything done here.

Timofffee commented:
Too bad, @Tribex. I really need this functionality now, because my application does not work correctly without it.

silverkorn commented:
https://github.com/godotengine/godot/pull/14622 was merged.
Would be great along with this PR!

hpvb commented:
The X11 code looks good. If the original author is gone I wouldn't mind bringing this up to date for master. It seems like a good feature combined with the new per-pixel transparency stuff.

Tribex commented:
@hpvb I'm still watching this thread, just don't have my build tooling or repo set up at the moment and haven't found time to do so.

mhilbrunner commented:
@hpvb Would love if you could take this over, would indeed be really nice with that one!

Timofffee commented:
Bump

Nobelix commented:
@hpvb Are you going to take over it ? Otherwise I'm okay with trying it to do

toger5 commented:
Really like the feature.
I'm curious, what is the benefit of this implementation compared to just exposing move/resize calls to the window manager.
Pseudocode:

Mouse relative movement event:
    If mouse over panel:
        Move window with mouse relative

toger5 commented:
Okay that makes a lot of sense. Just wanted to make sure it's a thought through decision!

silverkorn commented:
@Nobelix, according to what I understood, pretty much everything is there, @hpvb was mostly to make it fit with the latest code in the master... 
Maybe everyone stopped because of feature freeze and would prefer to make it to the next release?

@akien-mga, will it make it to 3.1 (As tagged) or it's preferred to wait for 3.2?

akien-mga commented:
> will it make it to 3.1 (As tagged) or it's preferred to wait for 3.2?

Might end up for 3.2 if it's not brought to a mergeable state before 3.1-beta.

toger5 commented:
This is also needed for wayland support #23209. Since wayland (xdg shell will not allow server side decoration).
I might do a rebase.

akien-mga commented:
Superseded by #23972.
